### PR TITLE
Feature/ejaku solution for RELEX challenge 

### PR DESCRIPTION
**Pull request by e-jaku**
This PR adds the necessary functionality for sending and parsing a csv file to a backend server.

Request are handled by the `CSVHandler` which takes care of extracting a passed CSV from the request.  Request body size is limited to at most 10 MB. Allowed content type is narrowed down to "text/csv".

For parsing and processing the CSV file, a `CSVParser` (implementing the required Parser interface by the handler) is provided. A file is parsed at a row by row base and processed concurrently.

The domain structure is defined as a `Node` hierarchy, and the CSVParser adds each row as a node to this hierarchy.

A Makefile is provided for working with the solution locally: 
```
make run
``` 

and  
```
make test
```

can be used to run the tests locally. 

For running the tests in the GitHub action, a different make target is provided (`make test-with-backend`) that spins up the server backend and runs the tests simultaneously. 

In the `/api`  an openapi spec of the provided API endpoint can be found.

### CODE REVIEW

File: io_test.go [hunk 1]
```diff
@@ -114,7 +113,7 @@ func Test(t *testing.T) {
 			}
 
 			if !reflect.DeepEqual(got, want) {
-				t.Errorf("got %v, want %v", got, test.Want)
+				t.Errorf("got %v, want %v", got, want)
```
Comment (furkancaglar) [inline io_test.go:116]: Praise nice fix 👏

**Review by furkancaglar — COMMENTED** [review]
Thank you for taking the time to complete the task!

Overall I like the solution! I have tried to do a review as close as much to a one I do at normal work PR, so please feel free to discuss and make amendments to the code as you see fit.

Some general questions:
- How would you make the code more production ready? What parts/aspects would you add?
- Imagine a scenario where we have to support very large files. In that case, if the request is taking too long how would you improve UX?

File: internal/csv/csv_parser.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	LV_PREFIX string = "level_"
+)
+
+type CSVParser struct {
+	concurrency int
+}
+
+func NewCSVParser(concurrency int) *CSVParser {
+	if concurrency < 1 {
+		concurrency = 10 // default to 10
```
Comment (stiganik) [inline internal/csv/csv_parser.go:28]: Argue for the usefulness of using concurrency for doing the csv to json encoding.

File: internal/csv/csv_parser.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+			row, err := csvReader.Read()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return xerrors.Errorf("failed to read CSV row: %w", err)
+			}
+			select {
+			case rowChan <- row:
+			case <-ctx.Done():
```
Comment (stiganik) [inline internal/csv/csv_parser.go:63]: Can you think of a good reason to change the order of these cases. Have ctx.Done before writing into the channel.

File: internal/csv/csv_parser.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+// extractHierarchyLevels builds the necessary levels contained in a row for adding to the node hierarchy.
+// If a row does contain less elements than specified in the header or if a parent element is empty while the child
+// element is present this method will return an error.
+// This method returns the constructed level hierarchy in increasing depth order f.e.
+// {"level_1", "level_2", "level_3"}
+func extractHierarchyLevels(row []string, colIndexes map[string]int) ([]string, error) {
+	levels := []string{}
+	currentParent := ""
+
+	for i := 1; i <= len(colIndexes)-1; i++ {
```
Comment (stiganik) [inline internal/csv/csv_parser.go:187]: If we had to support other schemas which may not conform to the "level_" standard, how would you change this function to support the new schema too

File: internal/domain/node.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+type Node struct {
+	mu       sync.Mutex       `json:"-"`
+	Item     bool             `json:"item,omitempty"`
+	Children map[string]*Node `json:"children,omitempty"`
+}
+
+func NewNode() *Node {
+	return &Node{
+		mu: sync.Mutex{},
```
Comment (stiganik) [inline internal/domain/node.go:13]: Do you think this would also work if you didn't create the mutex struct here ?

File: internal/server/server.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi/v5/middleware"
+)
+
+func NewServer(cfg *config.Config, f func(chi.Router)) *http.Server {
+	r := chi.NewRouter()
+
+	r.Use(middleware.RealIP)
+	r.Use(middleware.Recoverer)
+	r.Use(middleware.Timeout(cfg.RequestTimeout))
```
Comment (stiganik) [inline internal/server/server.go:16]: why not use the "Request" middleware which adds a requestID if the recoverer can make use of it ?

File: internal/server/server.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+	f(r)
+
+	return &http.Server{
+		Addr:         cfg.Address,
+		ReadTimeout:  cfg.ReadTimeout,
```
Comment (stiganik) [inline internal/server/server.go:22]: nice attention to detail

**Review by stiganik — COMMENTED** [review]
As a blanket question: why did you decide to keep the Go version as 1.12 rather than bumping it to 1.23 if it was a greenfield project?

File: internal/domain/node.go (hunk 5 above)
Comment (e-jaku) [reply internal/domain/node.go:13]: yes in this case it would be initialized automatically to its zero value since the outer struct (Node) is being initialized.

File: internal/csv/csv_parser.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+		select {
+		case row, ok := <-rowChan:
+			if !ok {
+				return nil
+			}
+
+			levels, err := extractHierarchyLevels(row, colIndexes)
+			if err != nil {
+				return xerrors.Errorf("failed to get hierarchy levels: %w", err)
+			}
```
[the remark this answers is not in the document]
Comment (e-jaku) [reply internal/csv/csv_parser.go:166]: ups, yup. Broke the use case while refactoring the solution to work with n levels. Added a check and tests covering it.

File: internal/csv/csv_parser.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+			itemIndex, ok := colIndexes[ITEM_ID]
+			if !ok || itemIndex >= len(row) {
+				return xerrors.Errorf("failed to parse row %v missing %s: %w", row, ITEM_ID, errors.ErrMissingRequiredValue)
+			}
+
+			nodes.AddNode(levels, row[itemIndex])
```
[the remark this answers is not in the document]
Comment (e-jaku) [reply internal/csv/csv_parser.go]: same as above, added check and test covering this edge case :)

File: internal/errors/errors.go [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+func IsKnownUserError(err error) bool {
+	for _, known := range KnownUserErrors {
+		if errors.Is(err, known) {
+			return true
+		}
+	}
+
+	return false
+}
```
[the remark this answers is not in the document]
Comment (e-jaku) [reply internal/errors/errors.go]: love the idea, wish I had it when implementing the challenge :smile:  avoids iterating and trying to unwrap multiple errors. Added the suggested change

File: internal/server/csv_handler.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+		r.Post("/", h.handleParseCSV)
+	})
+
+	return r
+}
+
+func (h *CSVHandler) handleParseCSV(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := h.logger.With().Str("handler", "handleParseCSV").Logger()
+	logger.Info().Msg("Parsing file...")
```
[the remark this answers is not in the document]
Comment (e-jaku) [reply internal/server/csv_handler.go:44]: Normally for different (bigger projects) I would prefer structured logging and build a process around log collecting and log analysis (like pushing logs for example to a Kafka queue and then to a logs consumer). For this particular small use case though, I would say nicely printed logs are my preferred choice (since I'm visually inspecting the logs and JSON formatted string are not so visually pleasing).

File: internal/server/csv_handler.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+			h.logger.Error().Err(err).Msg("Could not send json response")
+		}
+
+		return
+	}
+
+	if err := json.NewEncoder(w).Encode(&data); err != nil {
+		h.logger.Error().Err(err).Msg("Could not send json response")
+	}
+}
```
[the remark this answers is not in the document]
Comment (e-jaku) [reply internal/server/csv_handler.go]: Sure, valid suggestion, added change.  :)

File: cmd/backend/main.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+	done := make(chan struct{})
+	quit := make(chan os.Signal, 1)
+
+	csvParser := csv.NewCSVParser(20)
+
+	svr := server.NewServer(cfg, func(m chi.Router) {
+		m.Mount("/", server.NewCSVHandler(logger, csvParser).Router())
+	})
+
+	signal.Notify(quit, os.Interrupt)
```
[the remark this answers is not in the document]
Comment (e-jaku) [reply cmd/backend/main.go]: Sure, if it was deployed to Kubernetes would be such an example, there the SIGTERM signal is part of the pod lifecycle for shutting down.  Added. Locally I was only interrupting it via CTRL+c so it was not necessary in that scenario :)

File: internal/csv/csv_parser.go (hunk 3 above)
Comment (e-jaku) [reply internal/csv/csv_parser.go:62]: Good catch :man_facepalming: . Without moving it, it would probably not cancel if context is canceled, so long as the buffered channel is not full. Moved it to first position, nice one :+1:

File: internal/server/server.go (hunk 6 above)
Comment (e-jaku) [reply internal/server/server.go:17]: Strictly speaking not required, but I get your point. Makes a request more traceable. Added it

File: internal/csv/csv_parser.go (hunk 4 above)
Comment (e-jaku) [reply internal/csv/csv_parser.go:192]: Depends on the schema we want to support, I would say. As I interpreted the challenge, we are providing some hierarchical information with the numeration of the column names, but we could just extract the numerical value for example and use just that as a map key (without the prefix) that way we just access the levels directly and don't care about the prefix once the map is built.

File: internal/csv/csv_parser.go (hunk 2 above)
Comment (e-jaku) [reply internal/csv/csv_parser.go:28]: For this example, overkill. Even worse, for small to medium files, we might not even see performance improvements due to time wasted on context switching.  But on the other hand we have nicely separated rows which can be parsed simultaneously, we have no need to adhere to strict row order and when parsing the row we can make full use of the cpu resources available. Additionally, here we have quite a simple parsing/storing logic but if this was more complex or required async calls to different systems we for sure would see throughput gains in a concurrent processing.  Makes the solution more complex and difficult to maintain, though.

File: cmd/backend/main.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+	// ensure that the server shuts down gracefully
+	go func() {
```
[the remark this answers is not in the document]
Comment (e-jaku) [reply cmd/backend/main.go:45]: Probably readability and a simpler way of handling goroutine lifecycle by using the built-in context cancellation (for example by canceling the parent context on termination signal and triggering graceful termination of the service goroutines)

File: internal/csv/csv_parser.go (hunk 3 above)
Comment (stiganik) [reply internal/csv/csv_parser.go:62]: Actually the question was a red herring :D Golang documentation states that

> A select blocks until one of its cases can run, then it executes that case. It chooses one at random if multiple are ready.

So there is actually no difference in which order the cases are. I thought the same as you and had to look it up

Comment (e-jaku) [reply internal/csv/csv_parser.go:62]: Oh, good to know :) was not aware of the randomness behind it, seemed more logical that it would pick the first true statement
